Disassociating EIP when adding tags

### Terraform Version
Terraform v0.11.1
+ provider.aws v1.7.0
+ provider.null v1.0.0
+ provider.template v1.0.0

### Affected Resource(s)
- aws_eip

### Terraform Configuration Files
```hcl
resource "aws_eip" "nat-az-0" {
  vpc = true

  tags {
    Name        = "example"
  }
}

resource "aws_nat_gateway" "az-0" {
  allocation_id = "${aws_eip.nat-az-0.id}"
  subnet_id     = "${aws_subnet.dmz-az-0.id}"
}
```

### Debug Output
https://gist.github.com/williamlord/3618c58363da6a1383f4206b76ce047d

### Expected Behavior
Add tags.

### Actual Behavior
Attempted to disassociate EIP, AWS returned 400 Bad Request, 

### Steps to Reproduce
1. Create EIP without tags and NAT Gateway.
2. Add tags

### References
Attempting to use after https://github.com/terraform-providers/terraform-provider-aws/pull/2768.

I just replicated this too. Damn, I was so excited to be able to tag my EIPs, I downloaded 1.7.0 right away 😄 

Thank you @bflad for adding the tag support, and for fixing this bug so quickly. Hopefully 1.7.1 will be soon! 👏 

```
Terraform will perform the following actions:

  ~ module.nlb.aws_eip.subnet[0]
      tags.%:    "0" => "1"
      tags.Name: "" => "This doesn't work sad face"
```

```
* module.prod_nlb.aws_eip.subnet[0]: 1 error(s) occurred:

* aws_eip.subnet.0: AuthFailure: You do not have permission to access the specified resource.
        status code: 400, request id: 1234567-7932-4fc1-9d53-b6a87316abce
```

```
2018-01-12T15:21:22.580-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: Action=DisassociateAddress&AssociationId=eipassoc-8219b2a3&Version=2016-11-15
2018-01-12T15:21:22.580-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: -----------------------------------------------------
  tags.Name: "This works" => ""
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: 2018/01/12 15:21:22 [DEBUG] [aws-sdk-go] DEBUG: Response ec2/DisassociateAddress Details:
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: ---[ RESPONSE ]--------------------------------------
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: HTTP/1.1 400 Bad Request
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: Connection: close
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: Transfer-Encoding: chunked
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: Date: Fri, 12 Jan 2018 20:21:25 GMT
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: Server: AmazonEC2
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4:
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4:
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: -----------------------------------------------------
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: 2018/01/12 15:21:22 [DEBUG] [aws-sdk-go] <?xml version="1.0" encoding="UTF-8"?>
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: <Response><Errors><Error><Code>AuthFailure</Code><Message>You do not have permission to access the specified resource.</Message></Error></Errors><RequestID>1234567-7932-4fc1-9d53-b6a87316abce</RequestID></Response>
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4: 2018/01/12 15:21:22 [DEBUG] [aws-sdk-go] DEBUG: Validate Response ec2/DisassociateAddress failed, not retrying, error AuthFailure: You do not have permission to access the specified resource.
2018-01-12T15:21:22.781-0500 [DEBUG] plugin.terraform-provider-aws_v1.7.0_x4:   status code: 400, request id: 1234567-7932-4fc1-9d53-b6a87316abce
2018/01/12 15:21:22 [TRACE] root.prod_nlb: eval: *terraform.EvalWriteState
2018/01/12 15:21:22 [TRACE] root.prod_nlb: eval: *terraform.EvalApplyProvisioners
2018/01/12 15:21:22 [TRACE] root.prod_nlb: eval: *terraform.EvalIf
2018/01/12 15:21:22 [TRACE] root.prod_nlb: eval: *terraform.EvalWriteState
2018/01/12 15:21:22 [TRACE] root.prod_nlb: eval: *terraform.EvalWriteDiff
2018/01/12 15:21:22 [TRACE] root.prod_nlb: eval: *terraform.EvalApplyPost
2018/01/12 15:21:22 [ERROR] root.prod_nlb: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:
```

--- a/aws/resource_aws_eip.go
+++ b/aws/resource_aws_eip.go
@@ -225,17 +225,13 @@ func resourceAwsEipUpdate(d *schema.ResourceData, meta interface{}) error {
 
 	domain := resourceAwsEipDomain(d)
 
-	// Associate to instance or interface if specified
-	v_instance, ok_instance := d.GetOk("instance")
-	v_interface, ok_interface := d.GetOk("network_interface")
-
 	// If we are updating an EIP that is not newly created, and we are attached to
 	// an instance or interface, detach first.
 	disassociate := false
 	if !d.IsNewResource() {
-		if (d.Get("instance").(string) != "") && d.HasChange("instance") {
+		if d.HasChange("instance") && d.Get("instance").(string) != "" {
 			disassociate = true
-		} else if (d.Get("association_id").(string) != "") && (d.HasChange("network_interface") || d.HasChange("associate_with_private_ip")) {
+		} else if (d.HasChange("network_interface") || d.HasChange("associate_with_private_ip")) && d.Get("association_id").(string) != "" {
 			disassociate = true
 		}
 	}
@@ -245,10 +241,14 @@ func resourceAwsEipUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 	}
 
+	// Associate to instance or interface if specified
 	associate := false
-	if ok_instance && d.HasChange("instance") {
+	v_instance, ok_instance := d.GetOk("instance")
+	v_interface, ok_interface := d.GetOk("network_interface")
+
+	if d.HasChange("instance") && ok_instance {
 		associate = true
-	} else if ok_interface && (d.HasChange("network_interface") || d.HasChange("associate_with_private_ip")) {
+	} else if (d.HasChange("network_interface") || d.HasChange("associate_with_private_ip")) && ok_interface {
 		associate = true
 	}
 	if associate {

--- a/aws/resource_aws_eip.go
+++ b/aws/resource_aws_eip.go
@@ -231,13 +231,27 @@ func resourceAwsEipUpdate(d *schema.ResourceData, meta interface{}) error {
 
 	// If we are updating an EIP that is not newly created, and we are attached to
 	// an instance or interface, detach first.
-	if (d.Get("instance").(string) != "" || d.Get("association_id").(string) != "") && !d.IsNewResource() {
+	disassociate := false
+	if !d.IsNewResource() {
+		if (d.Get("instance").(string) != "") && d.HasChange("instance") {
+			disassociate = true
+		} else if (d.Get("association_id").(string) != "") && (d.HasChange("network_interface") || d.HasChange("associate_with_private_ip")) {
+			disassociate = true
+		}
+	}
+	if disassociate {
 		if err := disassociateEip(d, meta); err != nil {
 			return err
 		}
 	}
 
-	if ok_instance || ok_interface {
+	associate := false
+	if ok_instance && d.HasChange("instance") {
+		associate = true
+	} else if ok_interface && (d.HasChange("network_interface") || d.HasChange("associate_with_private_ip")) {
+		associate = true
+	}
+	if associate {
 		instanceId := v_instance.(string)
 		networkInterfaceId := v_interface.(string)
 

